registry/storage/driver/gcs/gcs.go

```diff
@@ -640,33 +652,29 @@ func (d *driver) Stat(context context.Context, path string) (storagedriver.FileI
 
 // List returns a list of the objects that are direct descendants of the
 // given path.
-func (d *driver) List(context context.Context, path string) ([]string, error) {
+func (d *driver) List(ctx context.Context, path string) ([]string, error) {
 	var query *storage.Query
 	query = &storage.Query{}
 	query.Delimiter = "/"
 	query.Prefix = d.pathToDirKey(path)
 	list := make([]string, 0, 64)
-	for {
-		objects, err := storageListObjects(d.context(context), d.bucket, query)
-		if err != nil {
-			return nil, err
-		}
-		for _, object := range objects.Results {
-			// GCS does not guarantee strong consistency between
-			// DELETE and LIST operations. Check that the object is not deleted,
-			// and filter out any objects with a non-zero time-deleted
-			if object.Deleted.IsZero() && object.ContentType != uploadSessionContentType {
-				list = append(list, d.keyToPath(object.Name))
-			}
+	objects, err := storageListObjects(ctx, d.bucket, query, d.gcs)
+	if err != nil {
+		return nil, err
+	}
+	for _, object := range objects {
+		// GCS does not guarantee strong consistency between
+		// DELETE and LIST operations. Check that the object is not deleted,
+		// and filter out any objects with a non-zero time-deleted
+		if object.Deleted.IsZero() && object.ContentType != uploadSessionContentType && object.Name != "" {
+
+			list = append(list, d.keyToPath(object.Name))
 		}
-		for _, subpath := range objects.Prefixes {
-			subpath = d.keyToPath(subpath)
+		if object.Name == "" && object.Prefix != "" {
+			subpath := d.keyToPath(object.Prefix)
```

Comment:
From a quick glance, it looks like in most cases we only need a limited set of the Object's attributes, but a lot more attributes are (potentially) available; https://pkg.go.dev/cloud.google.com/go/storage#ObjectAttrs

The documentation calls out that retrieving all attributes may have a performance impact; 
https://pkg.go.dev/cloud.google.com/go/storage#hdr-Listing_objects

> If only a subset of object attributes is needed when listing, specifying this subset using [Query.SetAttrSelection](https://pkg.go.dev/cloud.google.com/go/storage#Query.SetAttrSelection) may speed up the listing process:

Perhaps worth looking into if we can limit the attributes that we retrieve.